A75-xds-aggregate-cluster-behavior-fixes.md

```diff
@@ -0,0 +1,312 @@
+  xds_override_host policy assumes ownership of the subchannel instead
+  of deleting it.
+- In the picker, when we iterate through the addresses in the cookie, if
+  we do not find any entries with a subchannel is states READY, IDLE, or
+  CONNECTING and we do find entries that do not have any subchannel, we
+  will trigger creation of a subchannel for one of those entries.
+
+The picker logic will now look like this (pseudo-code):
+
+```
+def Pick(pick_args):
+  override_host_attribute = pick_args.call_attributes.get(attribute_key)
+  if override_host_attribute is not None:
+    entry_with_no_subchannel = None
+    idle_subchannel = None
+    found_connecting = False
+    for address in override_host_attribute.cookie_address_list:
+      entry = lb_policy.address_map[address]
+      if (entry found AND
+          entry.health_status is in policy_config.override_host_status):
+        if entry.subchannel is set:
+          if entry.subchannel.connectivity_state == READY:
+            override_host_attribute.set_actual_address_list(entry.address_list)
+            return entry.subchannel as pick result
+          elif entry.subchannel.connectivity_state == IDLE:
+            if idle_subchannel is None:
+              idle_subchannel = entry.subchannel
+          elif entry.subchannel.connectivity_state == CONNECTING:
+            found_connecting = True
+        else if entry_with_no_subchannel is None:
+          entry_with_no_subchannel = entry
+    # No READY subchannel found.  If we found an IDLE subchannel,
+    # trigger a connection attempt and queue the pick until that attempt
+    # completes.
+    if idle_subchannel is not None:
+      hop into control plane to trigger connection attempt for idle_subchannel
+      return queue as pick result
+    # No READY or IDLE subchannels.  If we found a CONNECTING
+    # subchannel, queue the pick and wait for the connection attempt
+    # to complete.
+    if found_connecting:
+      return queue as pick result
+    # No READY, IDLE, or CONNECTING subchannels.  If we found an entry
+    # with no subchannel, queue the pick and create a subchannel for
+    # that entry.
+    if entry_with_no_subchannel is not None:
+      hop into control plane to create subchannel for entry_with_no_subchannel
+      return queue as pick result
+  # pick_args.override_addresses not set or did not find a matching subchannel,
+  # so delegate to the child picker.
+  result = child_picker.Pick(pick_args)
+  if result.type == PICK_COMPLETE:
+    entry = lb_policy.address_map[result.subchannel.address()]
+    if entry found:
+      override_host_attribute.set_actual_address_list(entry.address_list)
+  return result
+```
+
+### Temporary environment variable protection
+
+This design does not provide any new functionality that will be enabled
+by external I/O, so no environment variable guard is necessary.
+
+## Rationale
+
+We considered implementing nested aggregate clusters as nested cds LB
```

Comment:
That's correct.  This flattening is not part of the proposal; it's an alternative that was considered but ruled out for the reasons explained here.

(Most gRFCs don't mention alternatives outside of the "rationale" section.  The dualstack gRFC was an exception to this, since everything was so tightly coupled that it was hard to describe the "what" without the "why", but it was an outlier.)

Reply:
This rationale section mentions two options: tree or flattening. We have to choose one or the other. Neither is mentioned in the proposal. It sounded like you wanted flattening:

> However, since the XdsDependencyManager already needs to resolve the
aggregate cluster dependency graph to make sure the entire config is
loaded, it seemed fairly straightforward to pass down the flattened list
and deal with it all in a single layer.

Reply:
The flattening isn't mentioned in the proposal, because it's not actually new in this proposal.  We were already doing it that way previously, so we're just sticking with that approach.

I've changed the wording here to try to clarify this.